src/base/UnorderedNonce.sol

```diff
@@ -1,53 +1,20 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 pragma solidity ^0.8.24;
 
+import {UnorderedNonceLibrary} from "../libraries/UnorderedNonceLibrary.sol";
+
 /// @title Unordered Nonce
 /// @notice Contract state and methods for using unordered nonces in signatures
 contract UnorderedNonce {
-    error NonceAlreadyUsed();
+    using UnorderedNonceLibrary for *;
 
     mapping(address owner => mapping(uint256 word => uint256 bitmap)) public nonces;
 
     /// @notice Consume a nonce, reverting if its already been used
     /// @param owner address, the owner/signer of the nonce
     /// @param nonce uint256, the nonce to consume
     function _useUnorderedNonce(address owner, uint256 nonce) internal {
-        // consume the bit by flipping it in storage
-        // reverts if the bit was already spent
-        _flipBit(nonces[owner], nonce);
-    }
-
-    function _flipBit(mapping(uint256 => uint256) storage bitmap, uint256 nonce) private {
-        // equivalent to:
-        //   uint256 wordPos = uint248(nonce >> 8);
-        //   uint256 bitPos = uint8(nonce);
-        //   uint256 bit = 1 << bitPos;
-        //   uint256 flipped = nonces[owner][wordPos] ^= bit;
-        //   if (flipped & bit == 0) revert NonceAlreadyUsed();
-
-        assembly ("memory-safe") {
-            // wordPos = uint248(nonce >> 8)
-            let wordPos := shr(8, nonce)
-
-            // bit = 1 << uint8(nonce)
-            let bit := shl(and(nonce, 0xFF), 1)
-
-            // slot of bitmap[wordPos] is keccak256(abi.encode(wordPos, bitmap.slot))
-            mstore(0, wordPos)
-            mstore(0x20, bitmap.slot)
-            let slot := keccak256(0, 0x40)
-
-            // uint256 previousBits = bitmap[wordPos]
-            let previousBits := sload(slot)
-
-            // revert if it's already been used
-            if eq(and(previousBits, bit), bit) {
-                mstore(0, 0x1fb09b80) // 4 bytes of NonceAlreadyUsed.selector
-                revert(0x1c, 0x04)
-            }
-
-            // bitmap[wordPos] = (previousBits | bit)
-            sstore(slot, or(previousBits, bit))
-        }
+        (uint256 wordPos, uint256 bitPos) = nonce.getBitmapPositions();
+        nonces[owner].flipBit(wordPos, bitPos);
```

Comment:
is it more expensive if we did nonces[owner][wordPos].flipBit(bitPos)? can you see? i think i just like that abstraction a little better

Reply:
its not as friendly since `flipBit` would calculate the new bitmap and return it so the caller can set it in storage

the alternative/cleaner syntax might be `nonces[owner].spendNonce(nonce)` where `spendNonce` does the wordPos/bitPos logic

Reply:
hmmm

Reply:
> nonces[owner][wordPos].flipBit(bitPos)

yeah generally nested mappings are much more expensive becaus eyou do more hashing (and hashing is on the verge of getting even more expensive, theres a change coming)

however i like `nonces[owner].spendNonce(nonce)` this!!

Reply:
ok i misunderstood - now i see it lol
i agree that doing it on the word would be messy becaus eyou have to write
```
nonces[owner][wordPos] = nonces[owner][wordPos].flipBit(x)
```

Reply:
i think i still prefer the abstraction where we were to change it to 

nonces[owner][wordPos].flipTick bc the functions become more reusable (to extract bitPos and wordPos) from nonce, then to get the new bitmap...

buttt i get it if its too expensive

maybe we could still have nonces[owner].spendNonce() that still use other helpers for those two things? im just noticing that we have to rewrite helpers in tests to do the same thing weve written here/ I could see those functions being helpful generally in a library

Reply:
Can you see how much gas is saved not doing it this way? @saucepoint 

like is it significantly more expsenive to do

nonces[owner][wordPos] = nonces[owner][wordPos].flipTick(bitPos) ?

Reply:
i will also say tho.. this is not a super heavily used codepath so we could just go back to the solidity version and be done w it, theres no need to hyper optimize this path

Reply:
yeah im happy with that 🤷 no strong opinions